### ISSUE TITLE
chain operator on develop branch

### ISSUE THREAD

**Issue by svlandeg** [explosion/thinc#208]
Something's wrong with the `chain` operator on the `develop` branch. 

This works:

```
layer1 = Linear(nO=9, nI=3)
layer2 = Linear(nO=12, nI=9)
layer3 = Linear(nO=5, nI=12)
model = chain(layer1, layer2, layer3)
model.set_dim("nO", 5)
```
But this doesn't : 
```
layer1 = Linear(nO=9, nI=3)
layer2 = Linear(nO=12, nI=9)
layer3 = Linear(nO=5, nI=12)
with Model.define_operators({">>": chain}):
    model = layer1 >> layer2 >> layer3
model.set_dim("nO", 5)
```
The latter crashes saying that the dimension of `nO` can't be changed from 12 to 5. It seems to do this because it's referring to `nO` of the SECOND layer, not the third, thus running into an inconsistency.

What seems to happen is that the second chain operator [breaks out of its constructor](https://github.com/explosion/thinc/blob/develop/thinc/layers/chain.py#L25) by extending `layer[0]` and returning that. But that seems to cause some trouble with the initialization flow....?

**Comment by honnibal** [comment]
Ah, hmm.

The chain operator's had a bunch of hacks that should be undone. The init logic is pretty horrible atm and doesn't work in lots of cases.

I think we should remove the claim that we can infer output sizes in chain. It'll let us unhack the init logic, and make things more consistent.

@ines Should we just make `nO` required instead of optional? It's bad that currently you can write a network with shapes that can't be inferred.

**Comment by svlandeg** [comment]
I'm not sure that the inference of output sizes really is the problem. I actually think that there's an internal inconsistency in the `self.layers` field. In my example above, I think it crashes because `model.layers[-1]` is pointing to the wrong layer, and if that's the case, training/prediction is probably going to fail, too ?

**Comment by honnibal** [comment]
Oh I see what you mean.

I bet it's an interaction with the operator overloading.

**Comment by svlandeg** [comment]
Note that this also doesn't work:

layer1 = Linear(nO=9, nI=3)
layer2 = Linear(nO=12, nI=9)
layer3 = Linear(nO=5, nI=12)
model1 = chain(layer1, layer2, layer3)

layer4 = Linear(nO=7, nI=5)
layer5 = Linear(nO=13, nI=7)
model2 = chain(layer4, layer5)

model = chain(model1, model2)
model.set_dim("nO", 13)

Now the final `model` seems to think it's `nO` is 5 (from `model1`) instead of 13. So it looks like this is an issue when combining multiple models (and in my original post, using `chain` created only one model with 3 layers, instead of this recursive adding of layers)

**Comment by honnibal** [comment]
We should just kill this auto-flattening case, it's Not Good. It will also lead to all sorts of problems with references.

I added it because for concatenate, it can be pretty inefficient to have nested concatenation instead of flat concatenation. But the same isn't true for chain, and the results can be surprising.

The specific bug happens because we set the `chain` `nO` dimension when we create the first chain object, then later we extend the outputs.
